### PR TITLE
Introduce a JGit extension

### PR DESCRIPTION
**Pull request by gastaldi**
This introduces a JGit extension, allowing to use [JGit](https://www.eclipse.org/jgit/) in a native image

### CODE REVIEW

File: extensions/jgit/deployment/pom.xml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-jgit-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>quarkus-jgit-deployment</artifactId>
+    <name>Quarkus - Jgit - Deployment</name>
```
Comment (gsmet) [inline extensions/jgit/deployment/pom.xml]:
```suggestion
    <name>Quarkus - JGit - Deployment</name>
```

File: extensions/jgit/deployment/src/main/java/io/quarkus/jgit/runtime/deployment/JGitProcessor.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ExtensionSslNativeSupportBuildItem;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.substrate.RuntimeInitializedClassBuildItem;
+import io.quarkus.deployment.builditem.substrate.SubstrateResourceBundleBuildItem;
+
+class JGitProcessor {
+
+    private static final String FEATURE = "jgit";
```
Comment (gsmet) [inline extensions/jgit/deployment/src/main/java/io/quarkus/jgit/runtime/deployment/JGitProcessor.java]: We will change it at some point but for now let's be consistent and add this to `FeatureBuildItem`. Please respect the alphabetical order there.

File: extensions/jgit/pom.xml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-extensions-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>quarkus-jgit-parent</artifactId>
+    <name>Quarkus - Jgit - Parent</name>
```
Comment (gsmet) [inline extensions/jgit/pom.xml]:
```suggestion
    <name>Quarkus - JGit - Parent</name>
```

File: extensions/jgit/runtime/pom.xml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-jgit-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>quarkus-jgit</artifactId>
+    <name>Quarkus - Jgit - Runtime</name>
```
Comment (gsmet) [inline extensions/jgit/runtime/pom.xml]:
```suggestion
    <name>Quarkus - JGit - Runtime</name>
```

**Comment by gastaldi** [comment]
As requested by @mdproctor

File: extensions/pom.xml [hunk 5]
```diff
@@ -107,6 +107,7 @@
 
         <!-- kogito -->
         <module>kogito</module>
+        <module>jgit</module>
```
Comment (gsmet) [inline extensions/pom.xml]: I would add it to the `Data access` section.

File: docs/src/main/asciidoc/jgit.adoc [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            return tmpDir.toString();
+        }
+    }
+
+----
+
+
+[WARNING]
+
+When running in native mode, make sure that https://quarkus.io/guides/native-and-ssl-guide#the-sunec-library-and-friends[SSL access is configured correctly]
```
Comment (gsmet) [inline docs/src/main/asciidoc/jgit.adoc]:
```suggestion
When running in native mode, make sure that https://quarkus.io/guides/native-and-ssl-guide#the-sunec-library-and-friends[SSL access is configured correctly].
```

File: docs/src/main/asciidoc/jgit.adoc [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        try (Git git = Git.cloneRepository().setDirectory(tmpDir).setURI(url).call()) {
+            return tmpDir.toString();
+        }
+    }
+
+----
+
+
+[WARNING]
+
```
Comment (gsmet) [inline docs/src/main/asciidoc/jgit.adoc]: Could you add the `====` to properly delimit the warning?

File: docs/src/main/asciidoc/jgit.adoc [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+Quarkus includes the `jgit` extension which enables the use of https://www.eclipse.org/jgit/[Eclipse JGit]
+in native mode.
+
+== Configuration
+
+Once you have your Quarkus project configured you can add the `jgit` extension
+to your project by running the following command in your project base directory.
+
+[source]
+mvn quarkus:add-extension -Dextensions="jgit"
```
Comment (gsmet) [inline docs/src/main/asciidoc/jgit.adoc]: I would prefer if you added the `----` to delimit the source extract.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a few very minor comments.

Interesting addition.

File: docs/src/main/asciidoc/jgit.adoc (hunk 8 above)
Comment (gastaldi) [reply docs/src/main/asciidoc/jgit.adoc]: Good catch. I copied from `kubernetes-client.adoc` :smile:

File: extensions/jgit/deployment/src/main/java/io/quarkus/jgit/runtime/deployment/JGitProcessor.java (hunk 2 above)
Comment (gastaldi) [reply extensions/jgit/deployment/src/main/java/io/quarkus/jgit/runtime/deployment/JGitProcessor.java]: Will do, that's what the `mvn quarkus:create-extension` outputs. Thanks for catching that!

Comment (gsmet) [reply extensions/jgit/deployment/src/main/java/io/quarkus/jgit/runtime/deployment/JGitProcessor.java]: Yeah, at some point, I think we will probably move them to each extension but better do it for all the extensions at once.

**Comment by gsmet** [comment]
@gastaldi I don't know exactly what's broken but all the tests fail in the Maven plugin tests.

Could you have a look?

**Comment by gastaldi** [comment]
Funny thing is that the tests pass when I run it locally :confused:

**Comment by gastaldi** [comment]
@gsmet found the problem: `quarkus-jgit-parent` was pointing to a wrong `<parent>`. For some reason that impacted the tests with `-Dnative`. It should be all good now :wink:

File: bom/deployment/pom.xml [hunk 9]
```diff
@@ -400,6 +400,11 @@
                 <artifactId>quarkus-integration-test-shared-library</artifactId>
                 <version>${project.version}</version>
             </dependency>
+            <dependency>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-jgit-deployment</artifactId>
+                <version>${project.version}</version>
+            </dependency>
```
Comment (gsmet) [inline bom/deployment/pom.xml]: It shouldn't be added there. These are the test dependencies section, it should be added above.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added one last minor comment. Could you address it?

Otherwise, it looks good to go.

**Comment by gastaldi** [comment]
@gsmet done, thanks for pointing that out!

File: integration-tests/jgit/src/test/java/io/quarkus/it/jgit/JGitTest.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    private final EmbeddedHttpGitServer gitServer = EmbeddedHttpGitServer
+            .fromBundle("booster-catalog", "repos/booster-catalog.bundle").usingAnyFreePort().create();
+
+    @BeforeEach
+    void startGitServer() throws Exception {
+        gitServer.start();
+    }
+
+    @Test
+    void should_clone() {
```
Comment (geoand) [inline integration-tests/jgit/src/test/java/io/quarkus/it/jgit/JGitTest.java]: Shouldn't camel case be used everywhere?

**Comment by geoand** [comment]
Cool addition 👌

File: integration-tests/jgit/src/test/java/io/quarkus/it/jgit/JGitTest.java (hunk 10 above)
Comment (geoand) [reply integration-tests/jgit/src/test/java/io/quarkus/it/jgit/JGitTest.java]: Could you please also squash to a single commit?

Comment (gastaldi) [reply integration-tests/jgit/src/test/java/io/quarkus/it/jgit/JGitTest.java]: I could, but isn't the big green Squash button enough? :smile:

Comment (geoand) [reply integration-tests/jgit/src/test/java/io/quarkus/it/jgit/JGitTest.java]: Not the way it is used in Quarkus 😎. In Quarkus a merge commit is always used, so committers are usually asked to squash there commits.
GitHub doesn't have a feature where you could squash commits and create a merge-commit as well unfortunately.

Comment (geoand) [reply integration-tests/jgit/src/test/java/io/quarkus/it/jgit/JGitTest.java]: Thanks!

**Comment by gsmet** [comment]
@gastaldi raah, sorry, we have a conflict now. Could you rebase (and fix the alphabetical order again?)?

Thanks!

**Review by gsmet — APPROVED** [review]

**Pull request merged**